include/deal.II/lac/petsc_ts.h

```diff
@@ -0,0 +1,593 @@
+      monitor;
+
+    /**
+     * Set up Jacobian callback without matrices.
+     *
+     * This callback gives full control to users to set up the linearized
+     * equations
+     * $\dfrac{\partial F}{\partial y} + \alpha \dfrac{\partial F}{\partial \dot
+     * y}$.
+     *
+     * All implicit solvers implementations are recast to use the above
+     * linearization. The $\alpha$ parameter is time-step and solver-type
+     * specific.
+     *
+     * Solvers must be provided via TimeStepper::solve_for_jacobian_system.
+     */
+    std::function<int(const real_type   t,
+                      const VectorType &y,
+                      const VectorType &ydot,
+                      const real_type   alpha)>
+      setup_jacobian;
+
+    /**
+     * Solution of the Jacobian system set up with TimeStepper::setup_jacobian.
+     *
+     * This is used as a preconditioner inside the Krylov process.
+     */
+    std::function<int(const VectorType &src, VectorType &dst)>
+      solve_for_jacobian_system;
+
+    /**
+     * Return an index set containing the algebraic components.
+     *
+     * Implementation of this function is optional. If your equation is also
+     * algebraic (i.e., it contains algebraic constraints, or Lagrange
+     * multipliers), you should implement this function in order to return only
+     * these components of your system.
+     */
+    std::function<IndexSet()> algebraic_components;
+
+  protected:
+    /**
+     * The PETSc object
+     */
+    TS ts;
+
+    SmartPointer<AMatrixType, TimeStepper> A;
+    SmartPointer<PMatrixType, TimeStepper> P;
+    bool                                   need_dae_tolerances;
+  };
+
+#  ifndef DOXYGEN
+  /*----------------------- Inline functions
+   * ----------------------------------*/
+
+  template <typename VectorType, typename PMatrixType, typename AMatrixType>
+  inline TS
+  TimeStepper<VectorType, PMatrixType, AMatrixType>::petsc_ts()
+  {
+    return ts;
+  }
+
+  template <typename VectorType, typename PMatrixType, typename AMatrixType>
+  inline MPI_Comm
+  TimeStepper<VectorType, PMatrixType, AMatrixType>::get_mpi_communicator()
+    const
+  {
+    return PetscObjectComm(reinterpret_cast<PetscObject>(ts));
+  }
+
+  template <typename VectorType, typename PMatrixType, typename AMatrixType>
+  inline typename TimeStepper<VectorType, PMatrixType, AMatrixType>::real_type
+  TimeStepper<VectorType, PMatrixType, AMatrixType>::get_time()
+  {
+    PetscReal      t;
+    PetscErrorCode ierr = TSGetTime(ts, &t);
+    AssertThrow(ierr == 0, ExcPETScError(ierr));
+    return t;
+  }
+
+  template <typename VectorType, typename PMatrixType, typename AMatrixType>
+  inline typename TimeStepper<VectorType, PMatrixType, AMatrixType>::real_type
+  TimeStepper<VectorType, PMatrixType, AMatrixType>::get_time_step()
+  {
+    PetscReal      dt;
+    PetscErrorCode ierr = TSGetTimeStep(ts, &dt);
+    AssertThrow(ierr == 0, ExcPETScError(ierr));
+    return dt;
+  }
```

Comment:
Not terribly important, but I don't think any of these functions are performance critical given what one would likely do with what they return. I would have put them into the `.cc` file as well.

Reply:
next time, please

Reply:
Certainly :-)